lib/charms/istio_k8s/v0/istio_info.py

```diff
@@ -0,0 +1,239 @@
+"""TODO: Add a proper docstring here.
+
+* using requirer:
+    * handle info changed event  <-- suggest a charm test to confirm this?
+
+* using provider:
+    * handle send error
+    * use the is_ready()
+
+
+requires cosl
+"""
+from typing import List, Optional
+
+from cosl.interfaces.utils import DatabagModelV2, DataValidationError
+from ops import CharmBase, Object, EventBase, CharmEvents, RelationDataAccessError, EventSource
+from pydantic import BaseModel, Field
+
+# The unique Charmhub library identifier, never change it
+LIBID = "17dff53d5bf649f29614365bde32451b"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+DEFAULT_RELATION_NAME = "istio-info"
+
+
+# Interface schema
+
+class IstioInfoAppData(DatabagModelV2, BaseModel):
+    """Data model for the istio-info interface."""
+
+    # TODO: Confirm that comparison between these objects works as expected so is_ready below works.
+
+    root_namespace: str = Field(
+        description="The root namespace for the Istio installation.",
+        examples=["istio-system"],
+    )
+
+    # To make it sortable, which is useful for comparing unordered sets like how relation data is stored.
+    # TODO: Move this to DatabagModelV2?
+    def __lt__(self, other):
+        return tuple(self.model_dump().values()) < tuple(other.model_dump().values())
+
+    def __le__(self, other):
+        return tuple(self.model_dump().values()) <= tuple(other.model_dump().values())
+
+    def __gt__(self, other):
+        return tuple(self.model_dump().values()) > tuple(other.model_dump().values())
+
+    def __ge__(self, other):
+        return tuple(self.model_dump().values()) >= tuple(other.model_dump().values())
+
+
+
+# Requirer library
+
+class DataChangedEvent(EventBase):
+    """Charm Event triggered when the istio-info relation changes."""
+
+
+class IstioInfoRequirerCharmEvents(CharmEvents):
+    """Events raised by the IstioInfoRequirer class."""
+    info_changed = EventSource(DataChangedEvent)
+
+
+class IstioInfoRequirer(Object):
+    """Class for handling the requirer side of the istio-info relation."""
+
+    on = IstioInfoRequirerCharmEvents()
+
+    def __init__(
+            self,
+            charm: CharmBase,
+            relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed,
+            self.on_relation_changed
+        )
+
+    def on_relation_changed(self, _: EventBase) -> None:
+        self.on.info_changed.emit()
+
+    def _get_relations(self):
+        """Return the applications related to us under the monitored relation."""
```

Comment:
I think this docstring should say "Return the relation instances" rather than "Return the applications" since it is in fact not returning applications.